File: Leetcode2906. Construct Product matrix/Solution.go
func constructProductMatrix(grid [][]int) [][]int {
	const MOD = 12345
	n, m := len(grid), len(grid[0])
	p := make([][]int, n)
	for i := range p {
		p[i] = make([]int, m)
	}

	suffix := int64(1)
	for i := n - 1; i >= 0; i-- {
		for j := m - 1; j >= 0; j-- {
			p[i][j] = int(suffix)
			suffix = (suffix * int64(grid[i][j])) % MOD
		}
	}

	prefix := int64(1)
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			p[i][j] = int((int64(p[i][j]) * prefix) % MOD)
			prefix = (prefix * int64(grid[i][j])) % MOD
		}
	}

	return p
}

File: Leetcode2906. Construct Product matrix/Solution.py
class Solution {
    public int[][] constructProductMatrix(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int size = m*n;
        int mod = 12345;

        int[] prefix = new int[size];
        int[] suffix = new int[size];
        int[] flat = new int[size];

        int index = 0;
        for(int[] row : grid){
            for(int val : row){
                flat[index++] = val % mod;
            }
        }

        prefix[0] = 1;
        for(int i=1;i<size;i++){
            prefix[i] = (prefix[i-1]*flat[i-1]) % mod;
        }

        suffix[size-1] = 1;
        for(int i= size -2;i>=0;i--){
            suffix[i] = (suffix[i+1] * flat[i+1]) % mod;
        }

        for(int i=0;i<size;i++){
            int val = (prefix[i] * suffix[i]) % mod;
            grid[i/n][i%n] = val;
        }

        return grid;

    }
}
